Rewrite template script for Alpine Linux

jirutka opened the pull request:
I’ve rewritten the Alpine template from scratch. The new template script is more readable and robust, uses cache and external LXC config file as other templates.

I hope that it can be added to lxc-ci now (https://github.com/lxc/lxc-ci/issues/4).

stgraber commented:
This pull request didn't trigger Jenkins as its author isn't in the whitelist.

An organization member must perform one of the following:

 - To have this branch tested by Jenkins, use the "ok to test" command.
 - To have a one time test done, use the "test this please" command.

Those commands are simple Github comments of the format: "jenkins: COMMAND"

jirutka commented:
@stgraber gentle ping

stgraber commented:
@ncopa Can you take a look?

ncopa commented:
Does it need to be bash instead of posix shell?

jirutka commented:
I'm not sure, I haven't tested it with other shell.

ncopa commented:
making it `#!/bin/sh` and fix the bashisms would make me alot happier, since we will not need to install bash for pure alpine setups (alpine build servers for example).

I would also prefer that we drop sys_adm by default like we did before. Alpine users tend to expect a more secure default, even if it means that some features won't work unless default config changes.

stgraber commented:
Marking as incomplete, I agree with @ncopa that we generally want to move our scripts away from bash, not add more dependencies on it.

jirutka commented:
Okay, now it runs even on the most minimal Alpine system without bash, rsync, curl or openssl. :smile_cat: I’ve also modified the default config to drop `sys_adm` capability.

And by the way, this is how it logs:

![screencast](https://cloud.githubusercontent.com/assets/949228/12531864/f1c79ad4-c205-11e5-8012-ee9ccceaf16a.gif)

kunkku commented:
One thing I would like to preserve from the current implementation is the possibility to specify additional packages to be installed as arguments.

The caching approach certainly speeds up the creation of the subsequent containers, but it has  drawbacks too. Firstly, any updates for packages will not be automatically checked. Secondly, the performance improvement would not apply to the additional packages specified on command line. My suggestion here is to leverage apk's package caching functionality instead.

I understand that many other templates use an approach similar to your implementation, but apk is an exceptionally fast package manager. Container creation with the base packages should not take longer than 5 seconds anyway on a decent machine when the packages have been cached. The problem in the current implementation is that the bootstrapping code does not set the cache up, leading to slower installation times on non-Alpine hosts.

jirutka commented:
> I would like to preserve from the current implementation is the possibility to specify additional packages to be installed as arguments.

In my opinion, build templates should do just one thing, build the base system, nothing more. It’s already quite complex script (349 LOC). If you need to install additional packages, then you can write a script just for this or use some configuration management tool. You would like to install just some additional packages, someone else would like to also configure this and that… I believe that this is not a good direction.

However, it’s not up to me to decide this.

> Firstly, any updates for packages will not be automatically checked.

Yes, obviously, but this is rather a feature – you will always get exactly the same image. If you want a fresh image, then just use `-F --flush-cache` option, e.g.:

    lxc-build -n alpine -t mycont -- -F

> …apk is an exceptionally fast package manager. Container creation with the base packages should not take longer than 5 seconds anyway…

That’s true, Alpine bootstrap is so fast that caching functionality is not so needed. However, what if you have some slow computer, say Raspberry Pi? I don’t know if it’s a real use case, but…

Anyway, the caching is almost “for free,” it doesn’t add much complexity to the code, so I think that we can keep it. Maybe we can switch it off by default, but I think that it’d be better to be consistent with other templates.

> …the bootstrapping code does not set the cache up, leading to slower installation times on non-Alpine hosts.

This is good idea, I’ll do it.

jirutka commented:
By the way, the primary, reason why I decided to rewrote the template, is to make it compatible with [lxc-ci](https://github.com/lxc/lxc-ci). This is currently needed to build an unprivileged container. I don’t know if it needs the caching functionality to work…?

stgraber commented:
lxc-ci doesn't need the cache feature, the way jenkins builds the images is by creation a container called "LXC_NAME", then tarring the result with some filtering and templating on top.

@ncopa can you take a look at the current state of this branch?

stgraber commented:
jenkins: test this please

lxc-jenkins commented:
This pull request didn't trigger Jenkins as its author isn't in the whitelist.

An organization member must perform one of the following:

 - To have this branch tested by Jenkins, use the "ok to test" command.
 - To have a one time test done, use the "test this please" command.

Those commands are simple Github comments of the format: "jenkins: COMMAND"

ncopa commented:
I agree with @kunkku wrt adding extra packages as arguments. I think people use this already so to avoid breakage for those I'd like to keep that functionality.

There are still various bashisms in the script. I doubt it will run at all on a debian/ubuntu host.

````console
$ checkbashisms lxc-alpine
possible bashism in lxc-alpine line 32 (export foo=bar should be foo=bar; export foo):
export PATH=$PATH:/usr/sbin:/usr/bin:/sbin:/bin
possible bashism in lxc-alpine line 50 (should be >word 2>&1):
if ! ls "$APK_KEYS_DIR"/alpine* &>/dev/null; then
possible bashism in lxc-alpine line 93 (echo -e):
	echo -e "ERROR: $@" >&2
possible bashism in lxc-alpine line 98 (echo -e):
	echo -e "\n==> $@"
possible bashism in lxc-alpine line 157 (alternative test command ([[ foo ]] should be [ foo ])):
	if [[ "$FLUSH_CACHE" = 'yes' && -d "$LXC_CACHE_DIR/bootstrap" ]]; then
possible bashism in lxc-alpine line 232 (should be 'b = a'):
	if [[ "$FLUSH_CACHE" == 'yes' && -d "$cache_dir" ]]; then
possible bashism in lxc-alpine line 232 (alternative test command ([[ foo ]] should be [ foo ])):
	if [[ "$FLUSH_CACHE" == 'yes' && -d "$cache_dir" ]]; then
possible bashism in lxc-alpine line 273 (should be >word 2>&1):
	cd - 1>&/dev/null
possible bashism in lxc-alpine line 481 (alternative test command ([[ foo ]] should be [ foo ])):
if [[ -z "$rootfs" && -f "$path/config" ]]; then
````

jirutka commented:
> There are still various bashisms in the script. I doubt it will run at all on a debian/ubuntu host.

Fixed and tested on ubuntu host.

> …the bootstrapping code does not set the cache up, leading to slower installation times on non-Alpine hosts.

@kunkku @ncopa How can I specify location of the local cache directory for APK? All I found is that APK expects symlink `/etc/apk/cache`, but I don’t like creating this in the host’s `/etc` (for non-Alpine hosts).

kunkku commented:
I think apk will use /etc of the specified root directory for this purpose.

ncopa commented:
I am ok to merge it if it does not break ubuntu hosts. We can do proper caching in later pull request.

brauner commented:
jenkins: test this please

kunkku commented:
@ncopa Merging the patches in their present state would remove the ability to install additional packages.

ncopa commented:
@kunkku oh. thats true.

@jirutka i think there are scripts out there in the wild using this feature. do you think you could add support for installing additional packages? I want avoid that stuff breaks for people...

jirutka commented:
I rewrote caching to cache APK packages instead of complete rootfs and added support for installing additional APK packages as requested. Sorry for long delay.

schu commented on templates/lxc-alpine.in:
```diff
@@ -1,367 +1,505 @@
+#=============================  Main  ==============================#
+
+if [ "$(id -u)" != "0" ]; then
+	die 1 "ERROR: This script must be run as 'root'"
```
This gives `ERROR: ERROR: This script must be run as 'root'`

jirutka replied:
Thanks, I’ve fixed it.

kunkku commented:
On which host distros has this now been tested?

jirutka commented:
Gentoo, Alpine and Ubuntu

kunkku commented:
In that case, I do not have any further objections to merging this.

How about you, @ncopa?

ncopa commented:
LGTM :+1:

@jirutka Good job!

brauner commented:
Good morning. :) I'll have Jenkins test this again before merging.

brauner commented:
jenkins: test this please

brauner commented:
Testing this manually here. When I try to create a normal root-owned container with `sudo lxc-create -n alpine -t alpine` with your patches applied I get the following error:

```
lxc-create: parse.c: lxc_file_for_each_line: 57 Failed to parse config: lxc.include = /usr/share/lxc/config/alpine.common.conf

lxc-create: lxc_create.c: main: 303 Error creating container alpine
```

brauner commented:
The error doesn't happen on current master and also not with lxc 1.1.5.

brauner commented:
Ok, I figured it out: You're missing entries to `Makefile.am` and `configure.ac`. If you don't include them `alpine.common.conf ` and `alpine.userns.conf` will not be installed.

```
---
 config/templates/Makefile.am | 2 ++
 configure.ac                 | 2 ++
 2 files changed, 4 insertions(+)

diff --git a/config/templates/Makefile.am b/config/templates/Makefile.am
index adc3faf..bf2f226 100644
--- a/config/templates/Makefile.am
+++ b/config/templates/Makefile.am
@@ -7,6 +7,8 @@ SUBDIRS = common.conf.d
 templatesconfig_DATA = \
 	archlinux.common.conf \
 	archlinux.userns.conf \
+	alpine.common.conf \
+	alpine.userns.conf \
 	centos.common.conf \
 	centos.userns.conf \
 	common.conf \
diff --git a/configure.ac b/configure.ac
index 6e4d0af..e3689ad 100644
--- a/configure.ac
+++ b/configure.ac
@@ -656,6 +656,8 @@ AC_CONFIG_FILES([
 	config/templates/Makefile
 	config/templates/archlinux.common.conf
 	config/templates/archlinux.userns.conf
+	config/templates/alpine.common.conf
+	config/templates/alpine.userns.conf
 	config/templates/centos.common.conf
 	config/templates/centos.userns.conf
 	config/templates/common.conf
-- 
2.7.1
```

brauner commented:
Feel free to use the patch snippet above. I waive all rights. :) If that's done I can merge.

jirutka commented:
Okay, I’ve fixed it.

brauner commented on templates/lxc-alpine.in:
```diff
@@ -1,367 +1,505 @@
+			shift; break
+		;;
+		--mapped-[ug]id)
+			die 1 "This template can't be used for unprivileged containers.\n" \
```
Two small things:

The `\n` will be printed literally in your `die()` function.

```
ERROR: This template can't be used for unprivileged containers.\n
ERROR: You may want to try the "download" template instead.
```

so you should probably remove it. :) Also, all other templates (well, nearly all), print:

```
This template can't be used for unprivileged containers.
You may want to try the "download" template instead.
```
without the `ERROR:` prepended. If you don't mind you could remove the `ERROR` for consistency among our scripts.

Otherwise, looks good! :)

jirutka replied:
I’ve removed `\n`, it was leftover from refactoring.

I don’t want to remove `ERROR: `, because it’s a good practice to highlight error messages like this to be better distinguishable from other output of the script.

brauner commented:
Thanks!
